test/e2e/re_inspection_test.go

```diff
@@ -0,0 +1,136 @@
+		Expect(err).NotTo(HaveOccurred())
+
+		By("creating a BMH with inspection disabled and hardware details added with wrong HostName")
+		newHardwareDetails := strings.Replace(hardwareDetails, "localhost.localdomain", wrongHostName, 1)
+		bmh := metal3api.BareMetalHost{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      specName + "-reinspect",
+				Namespace: namespace.Name,
+				Annotations: map[string]string{
+					metal3api.InspectAnnotationPrefix:   "disabled",
+					metal3api.HardwareDetailsAnnotation: newHardwareDetails,
+				},
+			},
+			Spec: metal3api.BareMetalHostSpec{
+				BMC: metal3api.BMCDetails{
+					Address:         bmcAddress,
+					CredentialsName: "bmc-credentials",
+				},
+				BootMode:       metal3api.Legacy,
+				BootMACAddress: bootMacAddress,
+			},
+		}
+		err = clusterProxy.GetClient().Create(ctx, &bmh)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("waiting for the BMH to be in registering state")
+		WaitForBmhInProvisioningState(ctx, WaitForBmhInProvisioningStateInput{
+			Client: clusterProxy.GetClient(),
+			Bmh:    bmh,
+			State:  metal3api.StateRegistering,
+		}, e2eConfig.GetIntervals(specName, "wait-registering")...)
+
+		By("waiting for the BMH to become available")
+		WaitForBmhInProvisioningState(ctx, WaitForBmhInProvisioningStateInput{
+			Client: clusterProxy.GetClient(),
+			Bmh:    bmh,
+			State:  metal3api.StateAvailable,
+		}, e2eConfig.GetIntervals(specName, "wait-available")...)
+
+		By("checking that the BMH has wrong HostName")
+		key := types.NamespacedName{Namespace: bmh.Namespace, Name: bmh.Name}
+		Expect(clusterProxy.GetClient().Get(ctx, key, &bmh)).To(Succeed())
+		Expect(bmh.Status.HardwareDetails.Hostname).To(Equal(wrongHostName))
+
+		By("removing HardwareDetailsAnnotation")
+		capm3_e2e.AnnotateBmh(ctx, clusterProxy.GetClient(), bmh, metal3api.HardwareDetailsAnnotation, nil)
+
+		By("adding InspectAnnotation to re-inspect")
+		capm3_e2e.AnnotateBmh(ctx, clusterProxy.GetClient(), bmh, metal3api.InspectAnnotationPrefix, pointer.String(""))
+
+		By("waiting for the BMH to be in inspecting state after inspection annotaion")
+		WaitForBmhInProvisioningState(ctx, WaitForBmhInProvisioningStateInput{
+			Client: clusterProxy.GetClient(),
+			Bmh:    bmh,
+			State:  metal3api.StateInspecting,
+		}, e2eConfig.GetIntervals(specName, "wait-inspecting")...)
+
+		By("waiting for the BMH to become available after re-inspection")
+		WaitForBmhInProvisioningState(ctx, WaitForBmhInProvisioningStateInput{
+			Client: clusterProxy.GetClient(),
+			Bmh:    bmh,
+			State:  metal3api.StateAvailable,
+		}, e2eConfig.GetIntervals(specName, "wait-available")...)
+
+		By("checking that the hardware details are corrected after re-inspection")
+		key = types.NamespacedName{Namespace: bmh.Namespace, Name: bmh.Name}
+		Expect(clusterProxy.GetClient().Get(ctx, key, &bmh)).To(Succeed())
+		Expect(bmh.Status.HardwareDetails.Hostname).To(Equal(rightHostName))
```

Comment:
We need to make this configurable so that the test can be used in other contexts. I think it will be fine to just add a variable to the e2e config with the expected hostname. To verify that it works, we can check with the fixture provider. The hardware details that it returns are hard coded here: https://github.com/metal3-io/baremetal-operator/blob/3874f1e2523335b50836712ff5597c297a48b1b6/pkg/provisioner/fixture/fixture.go#L136-L176

As you can see it does not set any hostname, so for fixture we would set `""`. If we can get the test passing on both fixture and ironic with separate e2e configs then it should be flexible enough for other environments also :slightly_smiling_face:

Reply:
added in e2e config, please check